python/tvm/contrib/pyfrontend.py

```diff
@@ -0,0 +1,305 @@
+        #assert isinstance(node.value, ast.Name) or isinstance(node.value, ast.Attribute)
+        args = self.visit(node.slice)
+        if isinstance(node.value, ast.Name):
+            array = node.value.id
+            assert array in self.input_param.keys()
+            _buf = self.input_param[array]
+            return _make.Call(_buf.dtype, array, args, _expr.Call.Halide, _buf.op, 0)
+        elif isinstance(node.value, ast.Attribute):
+            assert isinstance(node.value.value, ast.Name)
+            assert node.value.attr == "shape"
+            assert len(args) == 1
+            args = args[0]
+            #TODO: maybe support non-constant value later?
+            assert isinstance(args, _expr.IntImm) or isinstance(args, _expr.UIntImm)
+            assert node.value.value.id in self.input_param.keys()
+            return self.input_param[node.value.value.id].shape[args.value]
+        else:
+            assert False
+
+    def visit_With(self, node):
+        assert len(node.items) == 1
+        context = node.items[0].context_expr
+        option = node.items[0].optional_vars
+        assert isinstance(context, ast.Call)
+        assert isinstance(option, ast.Name)
+        self.annotation[option.id] = context.func.id
+        return _list_to_block([self.visit(i) for i in node.body])
+
+    def visit_BinOp(self, node):
+        lhs = self.visit(node.left)
+        rhs = self.visit(node.right)
+        return PyAST2HalideIR._binop_maker[type(node.op)](lhs, rhs)
+
+    def visit_For(self, node):
+        assert isinstance(node.target, ast.Name)
+        var = node.target.id
+        self.loop_levels[var] = _api.var(var)
+        assert isinstance(node.iter, ast.Call)
+        assert node.iter.func.id == 'range'
+        assert len(node.iter.args) == 1 or len(node.iter.args) == 2
+        if len(node.iter.args) == 1:
+            low, high = _api.const(0), self.visit(node.iter.args[0])
+        elif len(node.iter.args) == 2:
+            low, high = self.visit(node.iter.args[0]), self.visit(node.iter.args[1])
+        ext = high if isinstance(low, _expr.IntImm) and low.value == 0 else high - low
+        _body = [self.visit(stmt) for stmt in node.body]
+        _body = _list_to_block(_body)
+        _for_type = _stmt.For.Serial
+        if var in self.annotation.keys():
+            _for_type = eval('_stmt.For.%s' % self.annotation[var])
+        res = _make.For(self.loop_levels[var], low, ext, _for_type, 0, _body)
+        self.loop_levels.pop(var)
+        return res
+
+def parse(func, args):
+    """Parse a subset of Python to HalideIR
+
+    Parameters
+    ----------
+    func : str or types.FunctionType
+        If it is a string, parse the source code.
+        If it is a function, parse the function.
+
+    args : list of Buffer or Tensor or Var
+        The argument lists to the function.
+        Leave it None if no buffer is related to the function to be parsed
+
+    dump : bool, optional
```

Comment:
You may add it in parse above or just remove description here.

Reply:
OK. Let me do this.